File: LeetCode/List&LinkList/26.remove-duplicates-from-sorted-array.py
#
# @lc app=leetcode id=26 lang=python3
#
# [26] Remove Duplicates from Sorted Array
#

# @lc code=start
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        edit_p = 0
        explore_p = 0
        while (explore_p < len(nums)):
            if(nums[explore_p] > nums[edit_p]):
                nums[edit_p+1] = nums[explore_p]
                edit_p+=1
            explore_p+=1


        return edit_p+1

# 2个指针一个负责探索一个负责改
# 探索指针explore_p只要不碰到数组尾部，就不断往前走。只要发现比当前改指针edit_p大的数就修改改指针edit_p的下一个，并且把edit_p加1
# @lc code=end

File: LeetCode_Go/List&LinkedList/26.remove-duplicates-from-sorted-array.go
/*
 * @lc app=leetcode id=26 lang=golang
 *
 * [26] Remove Duplicates from Sorted Array
 */

// @lc code=start

// Remove duplicates in a sorted integer array in-place. Return the number of unique elements (k) and modify the array to include only the first k unique elements in sorted order.

// two pointers: curIdx, for populate the unique part; searchIdx, for traverse the array and search for next unique element. One state: curValue, store value of the last unique element, as a criterion for finding a next unique element.

// O(n)
func removeDuplicates(nums []int) int {
	if len(nums) <= 1 { // len 1 case
		return len(nums)
	}
	// len > 1 case
	curIdx, searchIdx := 1, 1
	curValue := nums[0]
	for searchIdx < len(nums) {
		if nums[searchIdx] <= curValue {
			searchIdx++
			continue
		}
		curValue = nums[searchIdx]
		nums[curIdx] = nums[searchIdx]
		curIdx++
	}
	return curIdx
}

// @lc code=end

